conn.go

```diff
@@ -978,6 +998,35 @@ func (c *Conn) exec(ctx context.Context, req frameWriter, tracer Tracer) (*frame
 	return framer, nil
 }
 
+// ObservedStream observes a single request/response stream.
+type ObservedStream struct {
+	// Host of the connection used to send the stream.
+	Host *HostInfo
+}
+
+// StreamObserver is notified about stream state.
+// A stream is started every time a request is written to the server
+// and is finished when a response is received.
+// While queries and streams rougly correspond to each other
+// (a single stream is used to get result of a query), a stream
+// might live longer than query (as reported by QueryObserver).
+// Specifically, when a query/batch times out, the QueryObserver/BatchObserver
+// report immediately, but the stream is still alive until either the response
+// is returned or the connection is closed.
+type StreamObserver interface {
+	// StreamStarted is called when a new stream is started.
+	// This happens just before a request is written to the wire.
+	StreamStarted(observedStream ObservedStream)
+
+	// StreamAbandoned is called when we stop waiting for response.
+	// This happens when a connection is closed.
+	// StreamFinished is won't be called for this stream.
+	StreamAbandoned(observedStream ObservedStream)
+
+	// StreamFinished is called when we receive a response for the stream.
+	StreamFinished(observedStream ObservedStream)
+}
```

Comment:
The majority of the other observers also take the context as a first parameter. Haven't we considered adding it too? It could help with monitoring by providing additional data to determine for which one call the query has been executed.

Ref:
- [QueryObserver](https://pkg.go.dev/github.com/gocql/gocql#QueryObserver)
- [FrameHeaderObserver](https://pkg.go.dev/github.com/gocql/gocql#FrameHeaderObserver)
- [BatchObserver](https://pkg.go.dev/github.com/gocql/gocql#BatchObserver)

Reply:
Good point!

We can add context to StreamStarted, but I'm not sure about StreamFinished/Abandoned - we'd need to store the context in the `call` to pass it to StreamFinished/Abandoned and those might be called long after the query finishes/context has been canceled. Perhaps we could allow StreamStarted to return an `interface{}` value that will be passed to StreamFinished/Abandoned instead of passing the context there?

Reply:
We can just add context to the [`conn.releaseStream()`](https://github.com/kiwicom/gocql/blob/ff7f4005/conn.go#L714-L714)as it's called only on three places:
- [In `(c *Conn) exec`](http://github.com/kiwicom/gocql/blob/ff7f4005d4f32bf5cc16ab73cda598d64901f0a2/conn.go#L965-L965)
- [In `(c *Conn) exec` again](http://github.com/kiwicom/gocql/blob/ff7f4005d4f32bf5cc16ab73cda598d64901f0a2/conn.go#L992-L992)
- [In `(c *Conn) recv`](http://github.com/kiwicom/gocql/blob/ff7f4005d4f32bf5cc16ab73cda598d64901f0a2/conn.go#L707-L707)

Ref: [sourcegraph search](https://sourcegraph.com/github.com/kiwicom/gocql@5ef205b1e9e0faededdb120c44e58ecd7d1a6ef0/-/blob/conn.go#L712:16&tab=references)

I'm also thinking if we need to add error to `StreamAbandoned` (not sure how many different errors could be passed via [`call.resp`](http://github.com/kiwicom/gocql/blob/ff7f4005d4f32bf5cc16ab73cda598d64901f0a2/conn.go#L957-L957))

Reply:
The context in `*Conn.recv` is not a real query context, it's based on `Session.ctx` (`Session.connect`→`Session.dial`→`Session.dialWithoutObserver`→`Conn.init`→`Conn.serve`→`Conn.recv`), so it does not have any useful info inside :shrug: 

StreamAbandoned is called when the connection that the stream was executed on is closed because of error, so I don't think we need to pass error there (we observe those using DisconnectObserver).

Reply:
I've updated the API, please check again.

Reply:
Yeah, didn't notice, that it has a different context 🙈